components/background_hang_monitor/sampler.rs

```diff
@@ -0,0 +1,337 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use backtrace;
+#[cfg(any(target_os = "android", target_os = "linux"))]
+use libc;
+#[cfg(target_os = "macos")]
+use mach;
+use msg::constellation_msg::{HangProfile, HangProfileSymbol};
+#[cfg(any(target_os = "android", target_os = "linux"))]
+use std::mem;
+use std::ptr;
+#[cfg(any(target_os = "android", target_os = "linux"))]
+use std::sync::atomic::{AtomicBool, Ordering};
+#[cfg(any(target_os = "android", target_os = "linux"))]
+use std::thread;
+#[cfg(any(target_os = "android", target_os = "linux"))]
+use std::time::Instant;
+
+const MAX_NATIVE_FRAMES: usize = 1024;
+
+/// The means of communication between monitored and monitor.
+#[cfg(any(target_os = "android", target_os = "linux"))]
+static mut SAMPLEE_CONTEXT: Option<(MonitoredThreadId, libc::mcontext_t)> = None;
```

Comment:
I've tried using the "atomic" crate, and it turned out that, on my Mac at least, those would end up using locks internally(there is a `Atomic::<T>::is_lock_free()` function that returned true on my machine). 

So the idea was instead to use a set of `AtomicBool` flags to regulate access to this unlocked data. It looks like those flags will now be replaced by a statemachine encoded in a `AtomicInt`...

Reply:
The issue is that without atomic reads/writes, you have no guarantee that the sampler thread will see the write made by the samplee.

Also, the `atomic` crate appears to provide atomic access to an entire type, instead of a pointer to it. I think the [atom](https://crates.io/crates/atom) would work well - it's a thin wrapper that allows you to easily use a `Box`ed type with `AtomicPtr`.

Reply:
Thanks, I will look into the "atom" crate. 

The goal is to implement similar to https://dxr.mozilla.org/mozilla-central/source/tools/profiler/core/platform-linux-android.cpp#148

which indeed seems to use a pointer to copy across the register state, see https://dxr.mozilla.org/mozilla-central/source/tools/profiler/core/platform-linux-android.cpp#215

Reply:
I think this still suffers from the same problem - since it's not accessed atomically, the sampler thread may not see writes made by the samplee thread.

Reply:
@jdm what do you think? 

I think the state machine, and the fact that this back and forth is limited to two threads, indirectly guarantees it's own kind of "atomicity".

Reply:
I think Aaron1011 has a point. The thread could perform the non-atomic write in such a way that when the other thread tries to read the value, it sees a different value instead: https://preshing.com/20130618/atomic-vs-non-atomic-operations/

Reply:
Thanks for the article. Could our use of `Ordering::SeqCst` prevent a data race problem? Having just [read more about it](https://doc.rust-lang.org/nomicon/atomics.html#sequentially-consistent), it appears to me that the read/write to the non-atomic data is "safe" thanks to the following:

1. Monitor loops and waits for `ThreadProfilerState::ReadyToObtainBacktrace` using `Ordering::SeqCst`.
2. Samplee writes the data. 
3. Samplee sets `ThreadProfilerState::ReadyToObtainBacktrace`, also using `Ordering::SeqCst`.
4. Monitor reads the data. 

From what I understand about "SeqCst access", is that it would prevent 2 and 4 above, which are indeed non-atomic in themselves, to be re-ordered in any way so that they would not occur in the order you would expect from ready the code. 

Please let me know what you think of this logic...

Reply:
Here is another related discussion: https://stackoverflow.com/questions/35153905/least-restrictive-memory-ordering-for-spin-lock-with-two-atomics?rq=1

I think we're fine. If anything, a future optimization might want to switch to aquire/release ordering as opposed to doing everything in `Ordering::SeqCst`.

What do you think?

Reply:
I've tried to make the workflow clearer with https://github.com/servo/servo/pull/21673/commits/8e65cf16e1abbce71cd2280d40e0e94fc604071c

Reply:
Unfortunately, I think that part of the documentation is incorrect - there's currently an open PR to change it: https://github.com/rust-lang/rust/pull/56023. `SeqCst` only prevent other `SeqCst` operations from being reordered - it doesn't affect non-atomic loads.

See also the [LLVM docs](https://llvm.org/docs/Atomics.html#sequentiallyconsistent) - `SeqCst` only provides guarantees about atomic operations.

Reply:
Ok thanks for pointing this out. If the "sequential" part only applies to the actual `load` and `store`, and not to whatever code happens to be sandwitched in between them, then I agree we'll need to wrap the shared data into it's own variant of atomic(maybe an `AtomicPtr` or indeed using the [`atom`](https://docs.rs/atom/0.3.5/atom/) crate. 

My impression from reading about memory ordering was instead that if you used `SeqCst`, it creates a kind of barrier that also makes the surrounding code sequentially consistent. 

@stjepang If you have a moment, could you please take a look at the following:

1. We're operating inside a "critical section"(a thread is currently suspended), and cannot use locks or heap allocation. 
2. We have two threads needing to do a workflow inside this section, including writing and reading to a piece of shared data.
3. We have a piece of shared data: `static mut SAMPLEE_CONTEXT: Option<(MonitoredThreadId, libc::mcontext_t)> = None;`
4. We have a state-machine used to coordinate the overall worflow, as well as the access to the shared data: 
```
static ref SAMPLER_STATE: AtomicUsize = AtomicUsize::new(ThreadProfilerState::NotProfiling as usize);

enum ThreadProfilerState {
    NotProfiling = 0,
    ReadyToWriteContext,
    ReadyToObtainBacktrace,
    ObtainedBacktrace,
    ResumedThread,
}
```
5. The actual workflow is implemented by storing and reading the ` AtomicUsize`, with the reading thread looping until the ` AtomicUsize` is in a specific state. 
6. All the stores and loads are done using `SeqCst`.
7. The shared data is written to by one thread, then that same threads does a `store` on the state machine. The other thread is doing a `load`, and waiting for that state, and only reading the data after the state machine has switched into it. 
8. This commit hopefully hightlights the overall flow: 8e65cf1

Question: Could the (non-atomic) reading and writing to the (unlocked) shared state, when sandwhiched in between storing and loading an atomic using `SeqCst`, still result in unexpected behavior?

Reply:
@Aaron1011 I find this (un-merged) update to the docs to be consistent with my earlier statement: 

> An operation on an atomic variable can do three things:
* Make the atomic itself safe to use from multiple threads.
* Additionally, synchronize other (even non-atomic) memory. A store with [`Release`] ordering
forms a synchronization (unidirectional) edge with any thread that does a load with
[`Acquire`] *on the written value*.
*Finally, a [`SeqCst`] operation participates in a globally consistent time line 

(see https://github.com/rust-lang/rust/pull/55233/files#diff-9aa1eb154ee636dcdcbcb58bc4c6f6c1R182)

Reply:
Finally, just looking at the docs:

> SequentiallyConsistent (seq_cst in IR) provides Acquire semantics for loads and Release semantics for stores. Additionally, it guarantees that a total ordering exists between all SequentiallyConsistent operations.
 
and 

> Acquire provides a barrier of the sort necessary to acquire a lock to access other memory with normal loads and stores.

and 

> Release is similar to Acquire, but with a barrier of the sort necessary to release a lock.

finally compare the above with 

> Monotonic is the weakest level of atomicity that can be used in synchronization primitives, although it does not provide any general synchronization. It essentially guarantees that if you take all the operations affecting a specific address, a consistent ordering exists.

@Aaron1011 Could it be that what you have in mind is this "Monotonic" model? This indeed only synchronizes the specific address inside the actual atomic. 

For the other definitions, I read them as such at `SequentiallyConsistent`, `Acquire`, and `Release` can be used to effectively lock other memory addresses than the one "contained" inside the atomic...

Reply:
@gterzian 

> My impression from reading about memory ordering was instead that if you used SeqCst, it creates a kind of barrier that also makes the surrounding code sequentially consistent.

Like @Aaron1011 said, this is not true. Atomic `SeqCst` operations are weaker than `SeqCst` [fences](https://doc.rust-lang.org/nightly/std/sync/atomic/fn.fence.html) (that's the kind of barrier you probably meant).

> If you have a moment, could you please take a look at the following:

The steps you wrote seem correct. You could even use `Acquire` and `Release` instead of `SeqCst` (but don't have to).

Just to clarify something:

* Mutable shared data **must** be atomic. No exceptions.
* Immutable shared data doesn't need to be atomic.
* Non-shared data doesn't need to be atomic (even if it is mutable and/or sent to another thread).

So if you want to have a mutable shared `Option<(MonitoredThreadId, libc::mcontext_t)>`, you must mutate an atomic variable. You should wrap the data in a `Mutex`, `AtomicPtr`, `Atom`, or something similar.

Note that in `Mutex<T>`, the internal mutex state is mutable and shared, while the `T` inside is not shared. In `RwLock<T>`, the `T` is always either shared and immutable (during read locks) or non-shared and mutable (during write locks).

It's also important to say that in order to send any kind of non-atomic data to another thread, you should synchronize it through a `Release` write and an `Acquire` read.

So when you create a `(MonitoredThreadId, libc::mcontext_t)` and want to make it visible to another thread, you must *publish* it through a write to the atomic variable with `Release` ordering (or stronger). The other thread then must read it with `Acquire` ordering (or stronger) in order to properly *fetch* the *published* data.

Reply:
@stjepang thanks for the explanatio, and @Aaron1011 thanks for persistently pointing out this problem!

I'll remove this code, related to the linux implementation, for now, and revisit in another PR. I might ask for your opinion again when the time comes...